librz/main/rz-diff.c

```diff
@@ -1,1305 +1,1454 @@
-		bufa = slurp(&ro, &c, ro.file, &fsz);
-		sza = fsz;
-		if (!bufa) {
-			eprintf("rz-diff: Cannot open %s\n", rz_str_get_null(ro.file));
-			return 1;
-		}
-		bufb = slurp(&ro, &c, ro.file2, &fsz);
-		szb = fsz;
-		if (!bufb) {
-			eprintf("rz-diff: Cannot open: %s\n", rz_str_get_null(ro.file2));
-			free(bufa);
-			return 1;
-		}
-		if (sza != szb) {
-			eprintf("File size differs %" PFMT64u " vs %" PFMT64u "\n", (ut64)sza, (ut64)szb);
-		}
+	case DIFF_TYPE_CLASSES:
+		diff = rz_diff_classes_new(&dfile_a, &dfile_b, ctx->compare_addresses);
 		break;
-	}
+	case DIFF_TYPE_COMMAND:
+		diff = rz_diff_command_new(ctx);
+		break;
+	case DIFF_TYPE_ENTRIES:
+		diff = rz_diff_entries_new(&dfile_a, &dfile_b);
+		break;
+	case DIFF_TYPE_FIELDS:
+		diff = rz_diff_fields_new(&dfile_a, &dfile_b, ctx->compare_addresses);
+		break;
+	case DIFF_TYPE_FUNCTIONS:
+		diff = rz_diff_functions_new(ctx);
+		break;
+	case DIFF_TYPE_IMPORTS:
+		diff = rz_diff_imports_new(&dfile_a, &dfile_b);
+		break;
+	case DIFF_TYPE_LIBRARIES:
+		diff = rz_diff_libraries_new(&dfile_a, &dfile_b);
+		break;
+	case DIFF_TYPE_LINES:
+		diff = rz_diff_lines_new((const char *)a_buffer, (const char *)b_buffer, NULL);
+		break;
+	case DIFF_TYPE_SECTIONS:
+		diff = rz_diff_sections_new(&dfile_a, &dfile_b, ctx->compare_addresses);
+		break;
+	case DIFF_TYPE_STRINGS:
+		diff = rz_diff_strings_new(&dfile_a, &dfile_b, ctx->compare_addresses);
+		break;
+	case DIFF_TYPE_SYMBOLS:
+		diff = rz_diff_symbols_new(&dfile_a, &dfile_b, ctx->compare_addresses);
+		break;
+	default:
+		rz_diff_error("unknown type\n");
+		goto rz_diff_unified_files_bad;
 	}
 
-	// initialize RzCons
-	(void)rz_cons_new();
+	if (!diff) {
+		goto rz_diff_unified_files_bad;
+	}
 
-	switch (ro.mode) {
-	case MODE_COLSII:
-		if (!c && !rz_list_empty(ro.evals)) {
-			c = opencore(&ro, NULL);
+	if (ctx->mode == DIFF_MODE_JSON) {
+		PJ *pj = rz_diff_unified_json(diff, ctx->file_a, ctx->file_b, ctx->show_time);
+		if (!pj) {
+			goto rz_diff_unified_files_bad;
 		}
-		dump_cols_hexii(bufa, (int)sza, bufb, (int)szb, (rz_cons_get_size(NULL) > 112) ? 16 : 8);
-		break;
-	case MODE_COLS:
-		if (!c && !rz_list_empty(ro.evals)) {
-			c = opencore(&ro, NULL);
+		printf("%s\n", pj_string(pj));
+		pj_free(pj);
+	} else {
+		// DIFF_MODE_STANDARD & DIFF_MODE_QUIET
+		char *result = rz_diff_unified_text(diff, ctx->file_a, ctx->file_b, ctx->show_time, ctx->colors);
+		if (!result) {
+			goto rz_diff_unified_files_bad;
 		}
-		dump_cols(bufa, (int)sza, bufb, (int)szb, (rz_cons_get_size(NULL) > 112) ? 16 : 8);
+		puts(result);
+		free(result);
+	}
+
+	result = true;
+
+rz_diff_unified_files_bad:
+	if (ctx->type == DIFF_TYPE_FUNCTIONS) {
+		rz_list_free((RzList *)rz_diff_get_a(diff));
+		rz_list_free((RzList *)rz_diff_get_b(diff));
+	}
+
+	rz_diff_free(diff);
+	rz_diff_file_close(&dfile_a);
+	rz_diff_file_close(&dfile_b);
+	free(a_buffer);
+	free(b_buffer);
+	return result;
+}
+
+RZ_API int rz_main_rz_diff(int argc, const char **argv) {
```

Comment:
Maybe we should introduce `RZ_MAIN` to distinguish "main" functions from the rest of the API

Reply:
i do not think this is needed.